### ISSUE TITLE
Field not accessible for guest-role if within Repeater and field access rights applied

### ISSUE THREAD

**Issue by tiefenb** [processwire/processwire-issues#1392]
I have a FieldtypeText within a FieldtypeRepeater. Both are accessible in the frontend for the guest-role.

When I change the field access rights for the FieldtypeText to "View: guest (everyone)", the text of the field is not visible anymore when I am logged out. If I am logged-in with my superuser-role account, the text is visible.

I need this access-rights for ProcessGraphQL.

![image](https://user-images.githubusercontent.com/488163/120287000-1ca43e00-c2bf-11eb-993c-bda2da0cbb17.png)

Tested with PHP 7.4.0, 8.0.3
MySQL 5.4 and 8.0
ProcessWire: 3.0.179

**Comment by ryancramerdesign** [comment]
@tiefenb Access with repeater items is delegated to the page that owns the repeater field.  Any access control added on top of this is likely to be confusing, since the access to the repeater is still the responsibility of the owning page. Make sure that your repeater items are being accessed from the page that owns them (rather than loaded separately from it). I would also suggest removing the field access control, as that would require that the user also has the requested permission to the template the the field exists on, and pages using repeater templates are not viewable on their own. In any case, I'm sure there's a way to achieve the access control scenario you are looking for, but probably just need to achieve it in a different way. If you can tell me more about the code you are using to access the data, I may be able to provide some suggestions on a good way to get there.

**Comment by tiefenb** [comment]
I am using the field in a very common template like ```foreach($page->my_repeater as $item) { echo $item->my_field; }```.

But I want to move to ProcessGraphQL, which needs access rights enabled on a field basis.

During the migration process, I need the field in the old template and also in ProcessGraphQL. After the migration process, everything should be served via ProcessGraphQL and the problem doesn't matter.

Maybe @dadish has also an idea?

**Comment by dadish** [comment]
@tiefenb Setting access control to the repeater field itself should make all the subfields within that repeater accessible by the guest user. Or you need a granular access control over subfields inside repeater?

**Comment by tiefenb** [comment]
@dadish If I set the access control to the repeater field makes it appear in GraphQL but without the subfields in it. If I want the sub fields I have also to give them access control which leads into this bug in the old frontend code.

**Comment by dadish** [comment]
@tiefenb Unfortunately I don't have a simple solution for you on this.

One approach you might take is to intercept the access to those fields via hooks in the old setup and grant/deny access to those individual fields. There are many places you can do it. One that comes to mind is `Fieldtype::wakeupValue`. It could look something like this.
```php
<?php
wire()->addHookAfter('Fieldtype::wakeupValue', function ($event) {
  $page = $event->arguments(0);
  $field = $event->arguments(1);
  $value = $event->arguments(2);
  if (/* the value can be accessed */) {
    // do nothing
  } else {
    $event->return = $field->getBlankValue();
  }
});
```

Let me know if that's something that could solve your problem.

**Comment by ryancramerdesign** [comment]
@tiefenb If I understand correctly, you need to have access control enabled for the Repeater field, as well as all the subfields within it, in order for it to work with the GraphQL module. Assuming "guest" has "view" permission on all of those then I think that should still work okay in your template code `foreach()` example you mentioned earlier. But access control can also be contextual per template, so I would look for any "Overrides" settings on your repeater field when in Setup > Fields > your_repeater_field. I would do the same for those subfields, just to make sure there isn't some other kind of contextual access control on top of it. 

Next, I think another to try would be enabling access control for the template used by your repeater field and giving "guest" view access to it. Access is usually delegated to the page that owns the repeater field, but in this case, your access control scenario is a bit more complex and layers within, so I'm thinking may need to be applied to the actual repeater template. You should be able to find it in Setup > Fields after selecting the "Show system fields" filter. 

Lastly, it would be good to know if there are any other modules applying access control rules. In your admin while in `$config->debug=true` mode, click the "Debug Mode Tools" link at the bottom. It will open a panel, click the "Hooks" item in the panel, and it'll reveal all the hooked methods. In your browser do a CMD-F/CTRL-F search for "viewable" and "hasPermission" to see what other hooks might be involved. Typically the only one you'll see would be `PagePermissions::viewable()` but I'm curious if there are others in your case.

**Comment by tiefenb** [comment]
Hi @ryancramerdesign!

Thanks for your time to investigate to this topic.
Today I wanted to look into this again I saw that it works like expected now.

I checked it on 3 different projects and it works now. I am a little bit confused, since I did not change anything on the settings.
The only thing I did is upgrading processwire to the latest commit in dev branch.

Did something change in the core which could affect this?

**Comment by matjazpotocnik** [comment]
@tiefenb, does Ryan's and Nurgulys answer help? Can you close this issue or perhaps add a response?

**Comment by netcarver** [comment]
@tiefenb Hi Markus,  could you update this please.

**Comment by tiefenb** [comment]
Hi @ryancramerdesign,

I checked the "Overrides" in Setup > Fields > your_repeater_field and the access-rights are correct and the same as in the general field settings.

Now I checked the access rights of the template of my repeater field and gave "guest" view access to it. And it seems this is the solution. I get the subfields in GraphQL *and* in my classic frontend.

I will close this but one last question: Can this change has any negative side effect?
